File: js/login-token-util.js
const crypto = require('crypto');

function generateLoginToken(username, userAgent, nonce, secret) {
    const encodedUsername = Buffer.from(username).toString('base64url');
    const dynamicSecret = secret + encodedUsername + nonce;
    const trimmedUserAgent = userAgent.substring(0, Math.min(120, userAgent.length));

    const hmac = crypto.createHmac('sha256', dynamicSecret);
    hmac.update(trimmedUserAgent);
    const token = hmac.digest('base64url');

    return `${encodedUsername}.${nonce}.${token}`;
}

function verifyLoginToken(token, userAgent, secret) {
    const tokenParts = token.split('.');
    if (tokenParts.length !== 3) {
        return false;
    }
    const encodedUsername = tokenParts[0];
    const username = Buffer.from(encodedUsername, 'base64url').toString();
    const nonce = tokenParts[1]
    const tokenOnly = tokenParts[2];
    const regeneratedToken = generateLoginToken(username, userAgent, nonce, secret);

    return token === regeneratedToken;
}

module.exports = { generateLoginToken, verifyLoginToken };

File: go/login_token_util.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"strings"
)

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func generateLoginToken(username, userAgent, nonce, secret string) string {
    encoder := base64.URLEncoding.WithPadding(base64.NoPadding)
    encodedUsername := encoder.EncodeToString([]byte(username))
    dynamicSecret := secret + encodedUsername + nonce
    trimmedUserAgent := userAgent[:min(120, len(userAgent))]

    hmac := hmac.New(sha256.New, []byte(dynamicSecret))
    hmac.Write([]byte(trimmedUserAgent))
    token := encoder.EncodeToString(hmac.Sum(nil))

    return fmt.Sprintf("%s.%s.%s", encodedUsername, nonce, token)
}

func verifyLoginToken(token, userAgent, secret string) bool {
    encoder := base64.URLEncoding.WithPadding(base64.NoPadding)
    parts := strings.Split(token, ".")
    if len(parts) != 3 {
        return false
    }

    encodedUsername := parts[0]
    usernameBytes, err := encoder.DecodeString(encodedUsername)
    if err != nil {
        fmt.Println("Error decoding username:", err)
        return false
    }
    username := string(usernameBytes)

    nonce := parts[1]
    regeneratedToken := generateLoginToken(username, userAgent, nonce, secret)

    return token == regeneratedToken
}